lib/puppet/transaction/resource_harness.rb

```diff
@@ -283,7 +283,24 @@ def new_system_value(property, event, old_system_value)
       # We use instance_variable_get here because we want this process to bypass any
       # munging/unmunging or validation that the property might try to do, since those
       # operations may not be correctly implemented for custom types.
-      property.instance_variable_get(:@should)
+      # require 'pry'; binding.pry if property.name == :permissions
+      should_value = property.instance_variable_get(:@should)
+      # wrapping trick to ensure Enumerable
+      [should_value].flatten.each { |i| customize_yaml_generation(i) }
+      should_value
+    end
+  end
+
+  def customize_yaml_generation(instance, to_remove = ['provider'])
+    # require 'pry'; binding.pry if instance.class.name == 'Puppet::Type::Acl::Ace'
+    if !instance.respond_to?(:encode_with) && to_remove.any? { |n| instance.instance_variable_defined?("@#{n}") }
+      singleton_class = class << instance; self; end
```

Comment:
You laugh, but there is a precedent for this btw - https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/file.rb#L302-L317